### ISSUE TITLE
Issues with str alias

### ISSUE THREAD

**Issue by srittau** [microsoft/pyright#1864]
**Describe the bug**
pyright reports errors when using a `str` alias in function arguments, e.g. like this:

```python
# locale.pyi

_str = str

# ...
def normalize(localename: _str) -> _str: ...
```

It works correctly in return types and in union argument types. It also works correctly when using `from builtins import str as _str`.

**To Reproduce**
I had no luck trying to reproduce this locally (but I admittedly did not try very hard), but python/typeshed#5465 fails CI with this change.

This is uncritical for typeshed as we have a workaround, but I wanted to bring it to your attention.

**Comment by erictraut** [comment]
Here's what's happening in this case.

Within stubs, all type references are implicitly forward references regardless of whether quotes are used. This applies to type aliases as well as other type annotations. For example, the following is legitimate:

```python
TypeAlias1 = Class1

def func1(a: TypeAlias1) -> None: ...

class Class1: ...
```

In your example, 

When the symbol `str` in the expression `_str = str` is evaluated, the analyzer finds that `str` is declared in the same scope (further down in the file), so it uses that definition rather than falling back to the outer (builtins) scope.

I see two ways to work around this:

1. `from builtins import str as _str` (which is what is currently in typeshed's local.pyi)
2. `import builtins` and `_str = builtins.str`

I agree that the error message that pyright emits in this case is confusing. I'll look at improving that. But I think it is correct for it to flag this situation as an error.

**Comment by srittau** [comment]
Thank you for the explanation, I agree that pyright's behavior makes sense, although I wonder why it doesn't flag `Optional[_str]` or `_str` when used as return value. Anyway, I will add a comment to the import and we will continue using it as before.

**Comment by erictraut** [comment]
I agree that there's something odd going on here, and I'll investigate further. I partly understand it, but more investigation is needed. It has to do with the way pyright hands cycles in type aliases (which are required for recursive type alias definitions).

Here's a minimal repro case:
```python
# pyright: strict
_str = str
def delocalize(string: _str) -> _str: ...
def str(val: float) -> _str: ...
```

**Comment by erictraut** [comment]
I've improved the logic to detect and report illegal type alias recursion, such as the case above where a type alias refers to a function that uses the type alias to annotate its parameters or return type.

**Comment by erictraut** [comment]
This is included in pyright 1.1.145, which I just published. It will also be included in the next release of pylance.
